File: Android/app/src/main/java/com/felipearpa/tyche/profile/AvatarImage.kt
package com.felipearpa.tyche.profile

import android.content.ContentResolver
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ImageDecoder
import android.graphics.Matrix
import android.graphics.Paint
import android.net.Uri
import androidx.compose.ui.geometry.Rect
import java.io.ByteArrayOutputStream

/**
 * Decodes with EXIF orientation applied — camera captures included — which is what the crop
 * geometry expects. Very large sources are downscaled at decode time to bound memory; the cap
 * stays far above the 512-pixel output times the 5× zoom budget.
 */
fun decodeOrientedBitmap(contentResolver: ContentResolver, uri: Uri): Bitmap? =
    runCatching {
        ImageDecoder.decodeBitmap(ImageDecoder.createSource(contentResolver, uri)) { decoder, info, _ ->
            decoder.allocator = ImageDecoder.ALLOCATOR_SOFTWARE

            val longestSide = maxOf(info.size.width, info.size.height)
            if (longestSide > MAX_DECODE_SIDE) {
                val downscale = MAX_DECODE_SIDE.toFloat() / longestSide
                decoder.setTargetSize(
                    (info.size.width * downscale).toInt(),
                    (info.size.height * downscale).toInt(),
                )
            }
        }
    }.getOrNull()

private const val MAX_DECODE_SIDE = 4096

/** Renders the crop-rect region into a square avatar and encodes it as JPEG. */
fun Bitmap.avatarRender(
    cropRect: Rect,
    outputSide: Int = AvatarCrop.OUTPUT_SIDE.toInt(),
    quality: Int = 80,
): Pair<Bitmap, ByteArray> {
    val output = Bitmap.createBitmap(outputSide, outputSide, Bitmap.Config.ARGB_8888)
    val renderScale = outputSide / cropRect.width

    val matrix = Matrix().apply {
        postTranslate(-cropRect.left, -cropRect.top)
        postScale(renderScale, renderScale)
    }

    Canvas(output).drawBitmap(this, matrix, Paint(Paint.FILTER_BITMAP_FLAG))

    val stream = ByteArrayOutputStream()
    output.compress(Bitmap.CompressFormat.JPEG, quality, stream)
    return output to stream.toByteArray()
}

File: iOS/Tyche/Tyche/UI/Profile/AvatarImage.swift
import UIKit

extension UIImage {
    /// Redraws the image so EXIF orientation is baked in and one point equals one source pixel,
    /// which is what the crop geometry expects. Very large sources are downscaled to bound
    /// memory; the cap stays far above the 512-pixel output times the 5× zoom budget.
    func orientationNormalized(maxSide: CGFloat = 4096) -> UIImage {
        let pixelSize = CGSize(width: size.width * scale, height: size.height * scale)
        let downscale = min(1, maxSide / max(pixelSize.width, pixelSize.height))

        if imageOrientation == .up && scale == 1 && downscale == 1 {
            return self
        }

        let targetSize = CGSize(
            width: (pixelSize.width * downscale).rounded(),
            height: (pixelSize.height * downscale).rounded()
        )

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1

        return UIGraphicsImageRenderer(size: targetSize, format: format).image { _ in
            draw(in: CGRect(origin: .zero, size: targetSize))
        }
    }

    /// Renders the crop-rect region into a square avatar and encodes it as JPEG.
    func avatarRender(
        cropRect: CGRect,
        outputSide: CGFloat = AvatarCrop.outputSide,
        quality: CGFloat = 0.8
    ) -> (image: UIImage, jpegData: Data)? {
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1

        let outputSize = CGSize(width: outputSide, height: outputSide)
        let renderScale = outputSide / cropRect.width

        let rendered = UIGraphicsImageRenderer(size: outputSize, format: format).image { _ in
            draw(
                in: CGRect(
                    x: -cropRect.minX * renderScale,
                    y: -cropRect.minY * renderScale,
                    width: size.width * renderScale,
                    height: size.height * renderScale
                )
            )
        }

        guard let jpegData = rendered.jpegData(compressionQuality: quality) else { return nil }

        return (rendered, jpegData)
    }
}
